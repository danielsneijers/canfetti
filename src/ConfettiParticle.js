import { randomColor } from './utils/colors';
import {
  degreeToRadian,
  randomCircumference,
  randomNumberBetween0andX,
} from './utils/math';
import Vector from './models/Vector';

export default class ConfettiParticle {
  constructor(ctx, x = 0, y = 0) {
    this.sqWidth = 20;
    this.sqHeight = 20;
    this.rotationDegree = 0;
    this.substractWidth = true;
    this.deltaTime = 0.016;
    this.color = randomColor();
    this.gravity = new Vector(0, 9.8);

    this.ctx = ctx;
    this.position = new Vector(x, y);
    this.velocity = new Vector(
      Math.sin(randomCircumference()) * Math.random() * 500,
      Math.cos(randomCircumference()) * Math.random() * 500,
    );
  }

  get inViewport() {
    return this.position.y < this.ctx.canvas.scrollHeight;
  }

  draw() {
    if (this.inViewport) {
      const { translateX, translateY } = this.updateParticle();

      this.ctx.save();

      this.ctx.translate(translateX, translateY);
      this.ctx.rotate(degreeToRadian(this.rotationDegree));
      this.ctx.translate(-translateX, -translateY);

      this.ctx.fillStyle = this.color;
      this.ctx.fillRect(
        this.position.x,
        this.position.y,
        this.sqWidth,
        this.sqHeight,
      );

      this.ctx.restore();
    }
  }

  updateParticle() {
    this.updateRotation(randomNumberBetween0andX(15));
    this.updateSubtractWidth();
    this.updateSqWidth();

    return this.updatePosition();
  }

  updatePosition() {
    const nextPosition = new Vector(
      this.velocity.x * this.deltaTime,
      this.velocity.y * this.deltaTime,
    );

    this.velocity.add(this.gravity).multiply(0.99);
    this.position.add(nextPosition);

    return {
      translateX: this.sqWidth / 2 + this.position.x,
      translateY: this.sqHeight / 2 + this.position.y,
    };
  }

  updateRotation(increment) {
    this.rotationDegree += increment;
  }

  updateSubtractWidth() {
    if (this.sqWidth <= -20) {
      this.substractWidth = false;
    }

    if (this.sqWidth >= 20) {
      this.substractWidth = true;
    }
  }

  updateSqWidth() {
    this.sqWidth = this.substractWidth
      ? this.sqWidth - randomNumberBetween0andX(5)
      : this.sqWidth + randomNumberBetween0andX(5);
  }
}
