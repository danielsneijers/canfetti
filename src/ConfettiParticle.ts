import { randomColor } from './utils/colors';
import Vector from './models/Vector';

export default class ConfettiParticle {
  private sqWidth: number = 20;
  private sqHeight: number = 20;
  private rotationDegree: number = 0;
  private substract: boolean = true;
  private dt: number = 0.016;
  private color: string;

  private position: Vector;
  private velocity: Vector = new Vector(0, 1);
  private gravity: Vector = new Vector(0, 9.8);
  private force: Vector = new Vector(0, 0);

  constructor(
    public ctx: CanvasRenderingContext2D,
    x: number = 10,
    y: number = 10,
    velocity?: Vector,
  ) {
    const r = Math.random() * Math.PI * 2;
    this.position = new Vector(x, y);
    this.color = randomColor();

    if (velocity) {
      this.velocity = velocity;
    } else {
      this.velocity = new Vector(
        Math.sin(r) * Math.random() * 500,
        Math.cos(r) * Math.random() * 500,
      );
    }
  }

  get inViewport() {
    return this.position.y < this.ctx.canvas.height / 2;
  }

  private get rotation(): number {
    return this.rotationDegree * Math.PI / 180;
  }

  public draw(): void {
    if (!this.inViewport) {
      return;
    }

    const { translateX, translateY } = this.updatePosition();

    this.ctx.save();

    this.ctx.translate(translateX, translateY);
    this.ctx.rotate(this.rotation);
    this.ctx.translate(-translateX, -translateY);

    this.ctx.fillStyle = this.color;
    this.ctx.fillRect(
      this.position.x,
      this.position.y,
      this.sqWidth,
      this.sqHeight,
    );

    this.ctx.restore();
  }

  private updatePosition(): { translateX: number; translateY: number } {
    const forceX = this.force.x + this.gravity.x;
    const forceY = this.force.y + this.gravity.y;
    const increment = Math.floor(Math.random() * 5);

    this.force.x = 0;
    this.force.y = 0;

    this.velocity.x += forceX;
    this.velocity.y += forceY;
    this.position.x += this.velocity.x * this.dt;
    this.position.y += this.velocity.y * this.dt;
    this.velocity.multiply(0.99);

    this.rotationDegree = this.rotationDegree += increment;

    this.sqWidth = this.substract
      ? this.sqWidth - increment
      : this.sqWidth + increment;

    if (this.sqWidth <= -20 || this.sqWidth >= 20) {
      this.substract = !this.substract;
    }

    return {
      translateX: this.sqWidth / 2 + this.position.x,
      translateY: this.sqHeight / 2 + this.position.y,
    };
  }
}
