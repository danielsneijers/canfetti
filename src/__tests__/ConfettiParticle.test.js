import ConfettiParticle from '../ConfettiParticle';

jest.mock('../utils/colors.js');

const mockCtx = {
  save: jest.fn(),
  translate: jest.fn(),
  rotate: jest.fn(),
  fillStyle: '#fff',
  fillRect: jest.fn(),
  restore: jest.fn(),
  canvas: {
    scrollHeight: 100,
  },
};

describe('ConfettiParticle', () => {
  let particleInViewport;
  let particleOutViewport;

  beforeAll(() => {
    global.Math.random = jest.fn(() => 0.6522531586373328);
  });

  afterAll(() => {
    global.Math.random.mockRestore();
  });

  beforeEach(() => {
    particleInViewport = new ConfettiParticle(mockCtx, 58, 23);
    particleOutViewport = new ConfettiParticle(mockCtx, 78, 100);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('constructor', () => {
    it('initiates with ctx and default values', () => {
      const particle = new ConfettiParticle(mockCtx);
      expect(particle).toMatchSnapshot();
    });
  });

  describe('inViewport', () => {
    it('returns true if particle is in viewport', () => {
      const particle2 = new ConfettiParticle(mockCtx, 16, 99);

      expect(particleInViewport.inViewport).toBe(true);
      expect(particle2.inViewport).toBe(true);
    });

    it('returns false if particle is not in viewport', () => {
      const particle2 = new ConfettiParticle(mockCtx, 32, 264);

      expect(particleOutViewport.inViewport).toBe(false);
      expect(particle2.inViewport).toBe(false);
    });
  });

  describe('draw', () => {
    it('draws when particle is in viewport', () => {
      expect(mockCtx.translate).not.toHaveBeenCalled();
      expect(mockCtx.rotate).not.toHaveBeenCalled();
      expect(mockCtx.fillRect).not.toHaveBeenCalled();

      particleInViewport.draw();

      expect(mockCtx.translate).toHaveBeenCalled();
      expect(mockCtx.rotate).toHaveBeenCalled();
      expect(mockCtx.fillRect).toHaveBeenCalled();
    });

    it('does not draw when particle is out of viewport', () => {
      expect(mockCtx.translate).not.toHaveBeenCalled();
      expect(mockCtx.rotate).not.toHaveBeenCalled();
      expect(mockCtx.fillRect).not.toHaveBeenCalled();

      particleOutViewport.draw();

      expect(mockCtx.translate).not.toHaveBeenCalled();
      expect(mockCtx.rotate).not.toHaveBeenCalled();
      expect(mockCtx.fillRect).not.toHaveBeenCalled();
    });
  });
});
