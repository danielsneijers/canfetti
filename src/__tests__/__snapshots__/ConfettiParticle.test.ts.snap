// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfettiParticle constructor initiates with ctx and default values 1`] = `
ConfettiParticle {
  "color": "#abcdef",
  "ctx": Object {
    "canvas": Object {
      "scrollHeight": 100,
    },
    "fillRect": [Function],
    "fillStyle": "#fff",
    "restore": [Function],
    "rotate": [Function],
    "save": [Function],
    "translate": [Function],
  },
  "deltaTime": 0.016,
  "gravity": Vector {
    "x": 0,
    "y": 9.8,
  },
  "position": Vector {
    "x": 0,
    "y": 0,
  },
  "rotationDegree": 0,
  "sqHeight": 20,
  "sqWidth": 20,
  "substractWidth": true,
  "velocity": Vector {
    "x": -266.5292067987189,
    "y": -187.9380952902847,
  },
}
`;
